eclair-core/src/main/scala/fr/acinq/eclair/payment/PaymentPacket.scala

```diff
@@ -317,15 +245,41 @@ object OutgoingPaymentPacket {
   }
   // @formatter:on
 
-  /**
-   * Build the command to add an HTLC with the given final payload and using the provided hops.
-   *
-   * @return the command and the onion shared secrets (used to decrypt the error in case of payment failure)
-   */
-  def buildCommand(replyTo: ActorRef, upstream: Upstream, paymentHash: ByteVector32, hops: Seq[ChannelHop], finalPayload: FinalPayload): Try[(CMD_ADD_HTLC, Seq[(ByteVector32, PublicKey)])] = {
-    buildPaymentPacket(paymentHash, hops, finalPayload).map {
-      case (firstAmount, firstExpiry, onion) =>
-        CMD_ADD_HTLC(replyTo, firstAmount, paymentHash, firstExpiry, onion.packet, None, Origin.Hot(replyTo, upstream), commit = true) -> onion.sharedSecrets
+  /** Build an encrypted onion packet from onion payloads and node public keys. */
+  def buildOnion(packetPayloadLength: Int, payloads: Seq[NodePayload], associatedData: ByteVector32): Either[OutgoingPaymentError, Sphinx.PacketAndSecrets] = {
+    val sessionKey = randomKey()
+    val nodeIds = payloads.map(_.nodeId)
+    val payloadsBin = payloads
+      .map(p => PaymentOnionCodecs.perHopPayloadCodec.encode(p.payload.records))
+      .map {
+        case Attempt.Successful(bits) => bits.bytes
+        case Attempt.Failure(cause) => return Left(CannotCreateOnion(cause.message))
+      }
+    Sphinx.create(sessionKey, packetPayloadLength, nodeIds, payloadsBin, Some(associatedData)) match {
+      case Failure(f) => Left(CannotCreateOnion(f.getMessage))
+      case Success(packet) => Right(packet)
+    }
+  }
+
+  /** Iteratively build all the payloads for a payment relayed through channel hops. */
+  def buildPayloads(finalAmount: MilliSatoshi, finalExpiry: CltvExpiry, finalPayload: NodePayload, hops: Seq[ChannelHop]): PaymentPayloads = {
+    // We ignore the first hop since the route starts at our node.
+    hops.tail.reverse.foldLeft(PaymentPayloads(finalAmount, finalExpiry, Seq(finalPayload))) {
+      case (current, hop) =>
+        val payload = NodePayload(hop.nodeId, IntermediatePayload.ChannelRelay.Standard(hop.shortChannelId, current.amount, current.expiry))
+        PaymentPayloads(current.amount + hop.fee(current.amount), current.expiry + hop.cltvExpiryDelta, payload +: current.payloads)
+    }
+  }
+
+  /** Build the command to add an HTLC for the given recipient using the provided route. */
+  def buildOutgoingPayment(replyTo: ActorRef, upstream: Upstream, paymentHash: ByteVector32, route: Route, recipient: Recipient): Either[OutgoingPaymentError, OutgoingPaymentPacket] = {
+    val outgoingChannel = route.hops.head.shortChannelId
+    for {
+      payment <- recipient.buildPayloads(paymentHash, route)
```

Comment:
It's probably just a matter of naming, because conceptually the `buildPayloads` that is here belongs more to this file than to the recipient, it's really a packet construction low-level detail...I'll have another look at that.

Reply:
Actually you're completely right, I was confusing that one with `buildOnion`, I'll move that to the `Recipient`.

Reply:
Moved in https://github.com/ACINQ/eclair/pull/2480/commits/f7ec174629b3e19fce90a423ccf6ef93c2ffc86f